File: domain/report.ts
/**
 * domain/report.ts
 *
 * Report is the pure domain model for the human-reviewed, publishable
 * synthesis of an Investigation's findings. Per the Human Accountability
 * principle, a Report may only transition to `active` (published) status
 * once it has both an author and a reviewer recorded — the platform may
 * draft and organize a report, but it never publishes on its own authority.
 *
 * Mirrors contracts/domain/report.schema.json field-for-field.
 *
 * Domain Purity: standard runtime and sibling domain modules only.
 */

import {
  Status,
  freezeRefs,
  freezeTags,
  newId,
  now,
  type Id,
  type Metadata,
  type Provenance,
  type Reference,
  type Timestamp,
  EMPTY_METADATA,
} from "./common.js";
import { ClassificationLevel } from "./classification_level.js";
import { ConfidenceScore } from "./confidence_score.js";
import { DomainInvariantError } from "./common.js";

export interface ReportInit {
  readonly investigationRef: Reference;
  readonly caseRef: Reference;
  readonly title: string;
  readonly summary: string;
  readonly authoredBy: string;
  readonly reviewedBy?: string;
  readonly findingRefs?: ReadonlyArray<Reference>;
  readonly status?: Status;
  readonly classification?: ClassificationLevel;
  readonly confidence?: ConfidenceScore;
  readonly publishedAt?: Timestamp;
  readonly provenance: Provenance;
  readonly tags?: ReadonlyArray<string>;
  readonly metadata?: Metadata;
}

/** Report is an immutable, human-authored synthesis document scoped to one Investigation. */
export class Report {
  readonly id: Id;
  readonly version: string;
  readonly createdAt: Timestamp;
  readonly updatedAt: Timestamp;
  readonly status: Status;
  readonly classification: ClassificationLevel;
  readonly investigationRef: Reference;
  readonly caseRef: Reference;
  readonly title: string;
  readonly summary: string;
  readonly authoredBy: string;
  readonly reviewedBy?: string;
  readonly findingRefs: ReadonlyArray<Reference>;
  readonly confidence?: ConfidenceScore;
  readonly publishedAt?: Timestamp;
  readonly provenance: Provenance;
  readonly tags: ReadonlyArray<string>;
  readonly metadata: Metadata;

  private constructor(id: Id, version: string, createdAt: Timestamp, updatedAt: Timestamp, init: ReportInit) {
    if (init.authoredBy.trim().length === 0) {
      throw new DomainInvariantError("Report.authoredBy must not be empty: reports may never be authorless.");
    }
    if (init.status === Status.Active && init.reviewedBy === undefined) {
      throw new DomainInvariantError("Report cannot reach 'active' status without a recorded reviewedBy.");
    }
    this.id = id;
    this.version = version;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
    this.status = init.status ?? Status.Draft;
    this.classification = init.classification ?? ClassificationLevel.Confidential;
    this.investigationRef = init.investigationRef;
    this.caseRef = init.caseRef;
    this.title = init.title;
    this.summary = init.summary;
    this.authoredBy = init.authoredBy;
    this.reviewedBy = init.reviewedBy;
    this.findingRefs = freezeRefs(init.findingRefs);
    this.confidence = init.confidence;
    this.publishedAt = init.publishedAt;
    this.provenance = init.provenance;
    this.tags = freezeTags(init.tags);
    this.metadata = init.metadata ?? EMPTY_METADATA;
    Object.freeze(this);
  }

  static draft(init: Omit<ReportInit, "status" | "publishedAt" | "reviewedBy">): Report {
    const timestamp = now();
    return new Report(newId(), "1.0.0", timestamp, timestamp, { ...init, status: Status.Draft });
  }

  /** Returns a new, human-reviewed Report; does not publish it. */
  withReview(reviewedBy: string): Report {
    if (reviewedBy.trim().length === 0) {
      throw new DomainInvariantError("withReview() requires a non-empty reviewer identity.");
    }
    return new Report(this.id, this.version, this.createdAt, now(), {
      investigationRef: this.investigationRef,
      caseRef: this.caseRef,
      title: this.title,
      summary: this.summary,
      authoredBy: this.authoredBy,
      reviewedBy,
      findingRefs: this.findingRefs,
      status: Status.UnderReview,
      classification: this.classification,
      confidence: this.confidence,
      provenance: this.provenance,
      tags: this.tags,
      metadata: this.metadata,
    });
  }

  /** Returns a new, published Report. Requires that a reviewer was already recorded. */
  publish(publishedAt: Timestamp = now()): Report {
    if (this.reviewedBy === undefined) {
      throw new DomainInvariantError("Report cannot be published before withReview() records a human reviewer.");
    }
    return new Report(this.id, this.version, this.createdAt, now(), {
      investigationRef: this.investigationRef,
      caseRef: this.caseRef,
      title: this.title,
      summary: this.summary,
      authoredBy: this.authoredBy,
      reviewedBy: this.reviewedBy,
      findingRefs: this.findingRefs,
      status: Status.Active,
      classification: this.classification,
      confidence: this.confidence,
      publishedAt,
      provenance: this.provenance,
      tags: this.tags,
      metadata: this.metadata,
    });
  }
}

File: domain/report.py
"""
domain/report.py

Report domain model - human-reviewable synthesis of investigation findings.
Mirrors contracts/domain/report.schema.json field-for-field.
"""
from dataclasses import dataclass, field
from enum import Enum
from typing import Sequence

from .classification_level import ClassificationLevel, ClassificationLevelValue
from .confidence_score import ConfidenceScore, ConfidenceLevel, ConfidenceMethods
from .common import (



    EMPTY_METADATA,
    Metadata,
    Provenance,
    Reference,
    Status,
    new_id,
    utc_now,
)


class ReportType(str, Enum):
    """Types of reports."""
    SUMMARY = "summary"
    DETAILED = "detailed"
    EXECUTIVE = "executive"
    TECHNICAL = "technical"
    LEGAL = "legal"
    THREAT_INTELLIGENCE = "threat_intelligence"
    INVESTIGATION = "investigation"
    OTHER = "other"


@dataclass(frozen=True)
class ReportSection:
    """A section within a report."""
    section_id: str
    title: str
    content: str
    order: int


@dataclass(frozen=True)
class Report:
    """
    Report is a human-reviewable synthesis of investigation findings.
    
    Per the Human Accountability principle, every Report requires a
    recorded human reviewed_by before it can reach active (published) status.
    The platform assists; it never publishes a report on its own authority.
    """
    id: str
    version: str
    created_at: str
    updated_at: str
    report_type: ReportType
    title: str
    summary: str
    status: Status
    classification: ClassificationLevel
    provenance: Provenance
    case_ref: Reference | None = None
    investigation_ref: Reference | None = None
    finding_refs: frozenset = field(default_factory=frozenset)
    sections: tuple[ReportSection, ...] = field(default_factory=tuple)
    confidence: ConfidenceScore | None = None
    author: str | None = None
    reviewed_by: str | None = None
    reviewed_at: str | None = None
    published_at: str | None = None
    tags: frozenset = field(default_factory=frozenset)
    metadata: Metadata = field(default_factory=lambda: dict(EMPTY_METADATA))

    @classmethod
    def create(
        cls,
        report_type: ReportType,
        title: str,
        summary: str,
        provenance: Provenance,
        finding_refs: Sequence[Reference] | None = None,
        sections: Sequence[ReportSection] | None = None,
        case_ref: Reference | None = None,
        investigation_ref: Reference | None = None,
        status: Status = Status.DRAFT,
        classification: ClassificationLevelValue = ClassificationLevelValue.INTERNAL,
        confidence: ConfidenceScore | None = None,
        author: str | None = None,
        tags: Sequence[str] | None = None,
        metadata: Metadata | None = None,
    ) -> "Report":
        """Create a new Report."""
        timestamp = utc_now()
        return cls(
            id=new_id(),
            version="1.0.0",
            created_at=timestamp,
            updated_at=timestamp,
            report_type=report_type,
            title=title,
            summary=summary,
            status=status,
            classification=ClassificationLevel(classification),
            provenance=provenance,
            case_ref=case_ref,
            investigation_ref=investigation_ref,
            finding_refs=frozenset(finding_refs) if finding_refs else frozenset(),
            sections=tuple(sections) if sections else tuple(),
            confidence=confidence,
            author=author,
            reviewed_by=None,
            reviewed_at=None,
            published_at=None,
            tags=frozenset(tags) if tags else frozenset(),
            metadata=dict(metadata) if metadata else dict(EMPTY_METADATA),
        )

    def submit_for_review(self, author: str) -> "Report":
        """Return a new Report submitted for review."""
        if self.status != Status.DRAFT:
            raise ValueError("Only draft reports can be submitted for review.")
        return Report(
            id=self.id,
            version=self.version,
            created_at=self.created_at,
            updated_at=utc_now(),
            report_type=self.report_type,
            title=self.title,
            summary=self.summary,
            status=Status.UNDER_REVIEW,
            classification=self.classification,
            provenance=self.provenance,
            case_ref=self.case_ref,
            investigation_ref=self.investigation_ref,
            finding_refs=self.finding_refs,
            sections=self.sections,
            confidence=self.confidence,
            author=author,
            reviewed_by=None,
            reviewed_at=None,
            published_at=None,
            tags=self.tags,
            metadata=self.metadata,
        )

    def approve(self, reviewer: str) -> "Report":
        """Return a new Report approved by the reviewer."""
        if self.status != Status.UNDER_REVIEW:
            raise ValueError("Only reports under review can be approved.")
        return Report(
            id=self.id,
            version=self.version,
            created_at=self.created_at,
            updated_at=utc_now(),
            report_type=self.report_type,
            title=self.title,
            summary=self.summary,
            status=Status.ACTIVE,
            classification=self.classification,
            provenance=self.provenance,
            case_ref=self.case_ref,
            investigation_ref=self.investigation_ref,
            finding_refs=self.finding_refs,
            sections=self.sections,
            confidence=self.confidence,
            author=self.author,
            reviewed_by=reviewer,
            reviewed_at=utc_now(),
            published_at=utc_now(),
            tags=self.tags,
            metadata=self.metadata,
        )

    def requires_review(self) -> bool:
        """Returns True if this report requires human review before publication."""
        return self.reviewed_by is None


__all__ = [
    "Report",
    "ReportSection",
    "ReportType",
]
